mne_bids/utils.py

```diff
@@ -26,9 +27,249 @@
+                 check_empty_room=True):
+        if all(ii is None for ii in [subject, session, task,
+                                     acquisition, run, processing,
+                                     recording, space, prefix, suffix]):
+            raise ValueError("At least one parameter must be given.")
+
+        self.update(subject=subject, session=session, task=task,
+                    acquisition=acquisition, run=run, processing=processing,
+                    recording=recording, space=space, prefix=prefix,
+                    suffix=suffix)
+
+        # check the task/session of er basename
+        if subject == 'emptyroom' and check_empty_room:
+            _check_empty_room_basename(self)
+
+    @property
+    def entities(self):
+        """Return dictionary of the BIDS entities."""
+        return OrderedDict([
+            ('subject', self.subject),
+            ('session', self.session),
+            ('task', self.task),
+            ('acquisition', self.acquisition),
+            ('run', self.run),
+            ('processing', self.processing),
+            ('recording', self.recording),
+            ('space', self.space),
+            ('prefix', self.prefix),
+            ('suffix', self.suffix)
+        ])
+
+    def __str__(self):
+        """Return the string representation of the path."""
+        basename = []
+        for key, val in self.entities.items():
+            if key not in ('prefix', 'suffix') and \
+                    val is not None:
+                _check_key_val(key, val)
+                if key == 'subject':
+                    key = 'sub'
+                if key == 'session':
+                    key = 'ses'
+                if key == 'acquisition':
+                    key = 'acq'
+                if key == 'processing':
+                    key = 'proc'
+                if key == 'recording':
+                    key = 'rec'
+                basename.append('%s-%s' % (key, val))
+
+        if self.suffix is not None:
+            basename.append(self.suffix)
+
+        basename = '_'.join(basename)
+        if self.prefix is not None:
+            basename = op.join(self.prefix, basename)
+
+        return basename
+
+    def __repr__(self):
+        """Representation in the style of `pathlib.Path`."""
+        return f'{self.__class__.__name__}({str(self)})'
+
+    def __fspath__(self):
+        """Return the string representation for any fs functions."""
+        return str(self)
+
+    def __eq__(self, other):
+        """Compare str representations."""
+        return str(self) == str(other)
+
+    def __ne__(self, other):
+        """Compare str representations."""
+        return str(self) != str(other)
+
+    def copy(self):
+        """Copy the instance.
+
+        Returns
+        -------
+        bidspath : instance of BIDSPath
+            The copied bidspath.
+        """
+        return deepcopy(self)
+
+    def get_bids_fname(self, kind=None, bids_root=None, extension=None):
```

Comment:
Essentially copied version of `_make_bids_fname`.

Reply:
So, is `BIDSPath` by default not the full path? Why is that?

Reply:
If the `BIDSPath` object is the full path, then it's not really the `bids_basename` anymore, which doesn't seem required by the API right now though? 

The internals of `read_raw_bids` and `write_raw_bids` doesn't require the full fname. This function operates like `_make_bids_fname` and can "infer" the kind and the extension by searching down the bids_root. 

We can either:
1. Modify this to incorporate a full path at all times? Not sure what this should look like though.
2. Turn this into a private function?
3. Add some further edits to this function, but keeping the header as is?

Reply:
@adam2392 let's discuss this in a new issue. We have spent too long on this PR. Let's merge this.

Reply:
Not really sure how to articulate the issue you're thinking of... :| can you make it?

Reply:
reading this again I think @jasmainak has a point here. The fact that get_bids_fname returns BIDSPath and not a str is a code smell. It suggests that BIDSPath can be informed by bids_root and kind.

wdyt?

Reply:
Ah yeah I see; it does seem like it could be confusing. So would the suggested fix be to:

1. make it return a str?
or
2. add bids_root/kind to BIDSPath explicitly?

Reply:
started a separate issue @adam2392